logging/interceptor.go

```diff
@@ -39,6 +60,133 @@ func LogLevelInterceptor(defaultLevel logrus.Level) grpc.UnaryServerInterceptor
+
+		ctx = fillInterceptor(ctx, fields, logger, options, startTime)
+
+		err := invoker(ctx, method, req, reply, cc, opts...)
+		if err != nil {
+			fields[logrus.ErrorKey] = err
+		}
+
+		code := status.Code(err)
+		fields[DefaultGRPCCodeKey] = code.String()
+
+		levelLogf(
+			logrus.NewEntry(logger).WithFields(fields),
+			options.codeToLevel(code),
+			"finished unary call with code "+code.String())
+
+		return err
+	}
+}
+
+func UnaryServerInterceptor(logger *logrus.Logger, opts ...Option) grpc.UnaryServerInterceptor {
+	options := initOptions(opts)
+
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		startTime := time.Now()
+		fields := newLoggerFields(info.FullMethod, startTime, DefaultServerKindValue)
+		newCtx := newLoggerForCall(ctx, logrus.NewEntry(logger), fields)
+
+		newCtx = fillInterceptor(newCtx, fields, logger, options, startTime)
+
+		resp, err := handler(newCtx, req)
+		if err != nil {
+			fields[logrus.ErrorKey] = err
+		}
+
+		code := status.Code(err)
+		fields[DefaultGRPCCodeKey] = code.String()
+
+		levelLogf(
+			ctxlogrus.Extract(newCtx).WithFields(fields),
+			options.codeToLevel(code),
+			"finished unary call with code "+code.String())
+
+		return resp, err
+	}
+}
+
+func fillInterceptor(ctx context.Context, fields logrus.Fields, logger *logrus.Logger, options *options, start time.Time) context.Context {
+	durField, durVal := options.durationFunc(time.Since(start))
+	fields[durField] = durVal
+
+	ctx = addRequestIDField(ctx, fields, logger, options)
+	ctx, err := addAccountIDField(ctx, fields, logger, options)
+	if err != nil {
+		logger.Warn(err)
+	}
+
+	for _, v := range options.fields {
+		ctx, err = addCustomField(ctx, fields, logger, v)
+		if err != nil {
+			logger.Warn(err)
+		}
+	}
+
+	return ctx
+}
+
+func addRequestIDField(ctx context.Context, fields logrus.Fields, logger *logrus.Logger, o *options) context.Context {
+	reqID, exists := requestid.FromContext(ctx)
+	if !exists || reqID == "" {
+		reqID = uuid.New().String()
+	}
+
+	fields[DefaultRequestIDKey] = reqID
+
+	return metadata.AppendToOutgoingContext(ctx, DefaultRequestIDKey, reqID)
+}
+
+func addAccountIDField(ctx context.Context, fields logrus.Fields, logger *logrus.Logger, o *options) (context.Context, error) {
+	accountID, err := auth.GetAccountID(ctx, nil)
+	if err != nil {
+		return ctx, fmt.Errorf("Unable to get %q from context", DefaultAccountIDKey)
+	}
+
+	fields[DefaultAccountIDKey] = accountID
+
+	return metadata.AppendToOutgoingContext(ctx, auth.MultiTenancyField, accountID), err
+}
+
+func addCustomField(ctx context.Context, fields logrus.Fields, logger *logrus.Logger, customField string) (context.Context, error) {
+	field, err := auth.GetJWTField(ctx, customField, nil)
+	if err != nil {
+		return ctx, fmt.Errorf("Unable to get custom %q field from context", customField)
+	}
+
+	fields[customField] = field
```

Comment:
Can we add the same option to get CustomFields from headers as well, from my point of view it can be reasonable